src/stores/ScannerStore.js

```diff
@@ -15,57 +15,114 @@
+
+import { NETWORK_LIST, NetworkProtocols } from '../constants';
 import { saveTx } from '../util/db';
-import { brainWalletSign, decryptData, keccak, ethSign } from '../util/native';
+import { isAscii } from '../util/message';
+import { blake2s, brainWalletSign, decryptData, keccak, ethSign, substrateSign } from '../util/native';
 import transaction from '../util/transaction';
+import { constructDataFromBytes, asciiToHex } from '../util/decoders';
 import { Account } from './AccountsStore';
 
 type TXRequest = Object;
 
 type SignedTX = {
-  txRequest: TXRequest,
+  recipient: Account,
   sender: Account,
-  recipient: Account
+  txRequest: TXRequest,
 };
 
 type ScannerState = {
-  type: 'transaction' | 'message',
-  txRequest: TXRequest | null,
+  dataToSign: string,
+  isHash: boolean,
+  isOversized: boolean,
   message: string,
-  tx: Object,
-  sender: Account,
+  multipartData: any,
   recipient: Account,
-  dataToSign: string,
-  signedData: string,
   scanErrorMsg: string,
-  signedTxList: [SignedTX]
+  sender: Account,
+  signedData: string,
+  signedTxList: [SignedTX],
+  tx: Object,
+  txRequest: TXRequest | null,
+  type: 'transaction' | 'message',
+  unsignedData: any
 };
 
 const defaultState = {
-  type: null,
   busy: false,
-  txRequest: null,
+  isHash: false,
+  isOversized: false,
+  dataToSign: '',
   message: null,
-  sender: null,
+  multipartData: {},
   recipient: null,
-  tx: '',
-  dataToSign: '',
+  scanErrorMsg: '',
+  sender: null,
   signedData: '',
-  scanErrorMsg: ''
+  tx: '',
+  txRequest: null,
+  type: null,
+  unsignedData: {}
 };
 
 export default class ScannerStore extends Container<ScannerState> {
   state = defaultState;
 
-  async setData(data, accountsStore) {
-    console.log(data);
-    switch (data.action) {
+  async setUnsigned(data) {
+    this.setState({
+      unsignedData: JSON.parse(data)
+    });
+  }
+
+  async setParsedData(strippedData, accountsStore) {
+    const parsedData = await constructDataFromBytes(strippedData);
+    
+    if (parsedData.isMultipart) {
+      this.setPartData(parseData.frame, parsedData.frameCount, parseData.partData, accountsStore);
+      return;
+    }
+
+    this.setState({
+      unsignedData: parsedData
+    });
+  }
+
+  async setPartData(frame, frameCount, partData, accountsStore) {
+    if (partData[0] === new Uint8Array([0x00]) || partData[0] === new Uint8Array([0x7B])) {
+      // part_data for frame 0 MUST NOT begin with byte 00 or byte 7B.
+      throw new Error('Error decoding invalid part data.');
+    }
+
+    // we havne't filled all the frames yet
+    if (Object.keys(this.state.multipartData.length) < frameCount) {
+      const nextDataState = this.state.multipartData;
+      
+      nextDataState[frame] = partData;
+
+      this.setState({
+        multipartData: nextDataState
+      });
+    }
+
+    // all the frames are filled
+    if (Object.keys(this.state.multipartData.length) === frameCount) {
```

Comment:
before we have the code:
```
 nextDataState[frame] = partData;
```
so the `this.state.multipartData` is filled not always from the first element, for example, check out the following example:
<img width="325" alt="Screenshot 2019-08-22 at 15 22 54" src="https://user-images.githubusercontent.com/6014309/63518183-c8b4d480-c4f0-11e9-83d6-d234f941c26d.png">

Reply:
Good point, I'll fix that with #322. For now I'll comment out the multipart part for this PR while I work on getting some test QRs ready.